dask/dataframe/shuffle.py

```diff
@@ -508,92 +508,93 @@ def rearrange_by_column_tasks(
+                    i,
+                )
 
-        join = {  # concatenate those pieces together, with their friends
-            ("shuffle-join-" + token, stage, inp): (
-                _concat,
-                [
+                _concat_list.append(
                     (
-                        "shuffle-split-" + token,
+                        shuffle_split_name,
                         stage,
                         inp[stage - 1],
-                        insert(inp, stage - 1, j),
+                        insert(inp, stage - 1, i),
                     )
-                    for j in range(k)
-                ],
+                )
+
+            # concatenate those pieces together, with their friends
+            join[(shuffle_join_name, stage, inp)] = (
+                _concat,
+                _concat_list,
                 ignore_index,
             )
-            for inp in inputs
-        }
+
         groups.append(group)
         splits.append(split)
         joins.append(join)
 
-    end = {
-        ("shuffle-" + token, i): ("shuffle-join-" + token, stages, inp)
-        for i, inp in enumerate(inputs)
-    }
+        end[(shuffle_token, idx)] = (shuffle_join_name, stages, inp)
 
     dsk = toolz.merge(start, end, *(groups + splits + joins))
-    graph = HighLevelGraph.from_collections("shuffle-" + token, dsk, dependencies=[df])
-    df2 = DataFrame(graph, "shuffle-" + token, df, df.divisions)
+    graph = HighLevelGraph.from_collections(shuffle_token, dsk, dependencies=[df])
+    df2 = DataFrame(graph, shuffle_token, df, df.divisions)
 
     if npartitions is not None and npartitions != df.npartitions:
-        parts = [i % df.npartitions for i in range(npartitions)]
         token = tokenize(df2, npartitions)
 
+        repartition_group_token = "repartition-group-" + token
+
         dsk = {
-            ("repartition-group-" + token, i): (
-                shuffle_group_2,
-                k,
-                column,
-                ignore_index,
-            )
+            (repartition_group_token, i): (shuffle_group_2, k, column, ignore_index,)
             for i, k in enumerate(df2.__dask_keys__())
         }
+
+        repartition_get_name = "repartition-get-" + token
+
         for p in range(npartitions):
-            dsk[("repartition-get-" + token, p)] = (
+            dsk[(repartition_get_name, p)] = (
                 shuffle_group_get,
-                ("repartition-group-" + token, parts[p]),
+                (repartition_group_token, p % df.npartitions),
                 p,
             )
 
         graph2 = HighLevelGraph.from_collections(
-            "repartition-get-" + token, dsk, dependencies=[df2]
-        )
-        df3 = DataFrame(
-            graph2, "repartition-get-" + token, df2, [None] * (npartitions + 1)
+            repartition_get_name, dsk, dependencies=[df2]
         )
+        df3 = DataFrame(graph2, repartition_get_name, df2, [None] * (npartitions + 1))
```

Comment:
Just realizing since we can/do use `shuffle` on a `dask_cudf` DataFrame, this should probably be something like `df.__class__` rather than `DataFrame`.

Reply:
See the `new_dd_object` function commonly used in dask/dataframe/core.py

Reply:
Made the change 👍